### ISSUE TITLE
version 2.13.0 introduced changes breaking reading from two results at once

### ISSUE THREAD

**Issue by rutek** [laminas/laminas-db#210]
### BC Break Report

Running two queries through MySQL PDO adapter returns two Result's. When they are stored, only one can be read. 

|    Q        |   A
|------------ | ------
| Version     | 2.13.0, 2.13.1 @ PHP 7.4 @ Ubuntu Server 20.04 LTS

#### Summary

When user made two queries and stored their results, first one returns result but second one throws exception. Such change was not mentioned in changelog and there were only "PHP 8.1 fixes" so I suppose it's not desired change.

Example script:

```php
<?php

namespace Test;

use Laminas\Db\Adapter\Adapter;

require_once 'vendor/autoload.php';

$db = new Adapter([
    'driver'   => 'Pdo_Mysql',
    'host' => '{host}',
    'database' => '{db}',
    'username' => '{user}',
    'password' => '{password}',
    'charset'  => 'UTF8',
]);

$result = $db->query('SELECT 1 total', []);
$result2 = $db->query('SELECT 1 total', []);
var_dump($result->toArray());
var_dump($result2->toArray());
```

#### Previous behavior

Both Results could be read successfully on 2.12.0.

```
array(1) {
  [0]=>
  array(1) {
    ["total"]=>
    string(1) "1"
  }
}
array(1) {
  [0]=>
  array(1) {
    ["total"]=>
    string(1) "1"
  }
}
```

#### Current behavior

Only one Result can be read.

```
array(1) {
  [0]=>
  array(1) {
    ["total"]=>
    string(1) "1"
  }
}
PHP Fatal error:  Uncaught Laminas\Db\Adapter\Exception\RuntimeException: This result is a forward only result set, calling rewind() after moving forward is not supported in /project/path/vendor/laminas/laminas-db/src/Adapter/Driver/Pdo/Result.php:200
Stack trace:
#0 /project/path/vendor/laminas/laminas-db/src/ResultSet/AbstractResultSet.php(245): Laminas\Db\Adapter\Driver\Pdo\Result->rewind()
#1 /project/path/vendor/laminas/laminas-db/src/ResultSet/AbstractResultSet.php(281): Laminas\Db\ResultSet\AbstractResultSet->rewind()
#2 /project/path/test.php(21): Laminas\Db\ResultSet\AbstractResultSet->toArray()
#3 {main}
  thrown in /project/path/vendor/laminas/laminas-db/src/Adapter/Driver/Pdo/Result.php on line 200
```

#### How to reproduce

By running script attached above. It runs without errors on laminas-db 2.12.0 but throws exception on 2.13.x.

**Comment by Ocramius** [comment]
Could this perhaps be isolated to a smaller test case?

Relevant changes seem to be:

 * https://github.com/laminas/laminas-db/compare/2.12.0...2.13.1#diff-8b9863458b0ac65210ea095907fcc634f58bb21f7fbb1b7e3c270b1c63fade81R174
 * https://github.com/laminas/laminas-db/compare/2.12.0...2.13.1#diff-55955428038b9bf948e0383865ab55ceee44aabe8418040d92e9b78814064b0fR198

The above `clone` call seems to be broken/replaced incorrectly by CS tooling

**Comment by Ocramius** [comment]
I'd say that we need to do a deep dive of `clone` usages in all that diff: could you try and give it a shot.

**Comment by Ocramius** [comment]
See https://github.com/laminas/laminas-db/pull/211 - please do give me a :+1: if it fixes your issue, as I only had time to set up a unit test, not a full integration suite (since I don't have Vagrant in my dev environment)

**Comment by rutek** [comment]
@Ocramius It looks OK. I cloned your branch `Ocramius:fix/#210-correct-resultset-cloning-to-avoid-iterator-rewind-issues` and replaced it locally. Result is valid:

```
array(1) {
  [0]=>
  array(1) {
    ["total"]=>
    string(1) "1"
  }
}
array(1) {
  [0]=>
  array(1) {
    ["total"]=>
    string(1) "1"
  }
}
```

**Comment by Ocramius** [comment]
Thanks, rolling a release then 👍

**Comment by Ocramius** [comment]
Not yet fixed: will need to re-release, sorry

**Comment by rutek** [comment]
@Ocramius Thanks :D I updated laminas-db to 2.13.2 from Composer and it still throws this Exception. I think it's something related to package build process as bundle do 2.13.2 does not contain your changes:

```
diff --no-dereference laminas-db/src/Adapter/Adapter.php laminas-db-2.13.2/src/Adapter/Adapter.php
198,203c198,200
<             $resultSet     = $resultPrototype ?? $this->queryResultSetPrototype;
<             $resultSetCopy = clone $resultSet;
< 
<             $resultSetCopy->initialize($result);
< 
<             return $resultSetCopy;
---
>             $resultSet = $resultPrototype !== null ? clone $resultPrototype : $this->queryResultSetPrototype;
>             $resultSet->initialize($result);
>             return $resultSet;
```

This is comparison of your pull request's branch and source code ZIP attached to 2.13.2. I think that tag 2.13.2 was created before pull request has been merged as there is no difference between 2.13.1 and 2.13.2:

https://github.com/laminas/laminas-db/compare/2.13.1...2.13.2

**Comment by rutek** [comment]
Okay, I missed your second comment - I think you found it before me :D

**Comment by Ocramius** [comment]
Now fixed in `2.13.3`.

Had to sleep over it, too risky to operate while that sleepy :D 

Thanks for the :beer:, @rutek!

**Comment by rutek** [comment]
I can confirm, now it works perfectly at 2.13.3. I've checked it using test script posted above and within project I've found this problem and query results are the same as 2.12.x (so they are OK), thanks! :)
